### DISCUSSION TITLE
Is null-state static analysis half-done?

### DISCUSSION THREAD

**Discussion by deep-outcome** [dotnet/csharplang#4964] [General]
I was grubbing among such said _null-state static analysis_, for instance [Attributes for null-state static analysis](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/attributes/nullable-analysis) and I can tell some acumens so far:

1. Why it is _null-state analysis_ if it works with `nullable` `value types` which do not have `null` value but `default` value?

>It is well known that for `nullable` `value type` `null` assignation is equal to assignation of `default` (`default(Nullable<X>)`. (Drolly `reference type` has it just opposite.)

```c#
> int? i = null;
. Console.Write ("#1 "); Console.WriteLine (new Nullable<int> () == i);
. Console.Write ("#2 "); Console.WriteLine (new int? () == i);
. Console.Write ("#3 "); Console.WriteLine (default(int?) == i);
. Console.Write ("#4 "); Console.WriteLine (default(Nullable<int>) == i);
. Console.Write ("#5 "); Console.WriteLine (default == i);
#1 True
#2 True
#3 True
#4 True
#5 True
```
2. Practically the same goes for _null-state static analysis_ **attributes**. Why they use `null` keyword when in .NET environment is preferable to use `default` whenever possible. This goes even for case if **attributes** worked just with `reference type` (and they do not).
3. What can it fix when compiler generates just _warnings_ by default? It is expected that development leader will switch the rules to _errors_ for whole team?
4. It has imperfect documentation with some not pretty mistakes like claim that `nullable-annotation` of `reference type` is part of method signature.
5. It has strong tendency to present that `nullable` `value types` magically can have `null` value.

> Value type is never `null`. Even when it is declared as `nullable` via `?` suffix, it is just shorthand for _generic constructed closed value type_, `Nullable<X>`, where `X` is corresponding `value type`. Thus to state out that `null` is fine value for instance of `Nullable<int>` variable is not true at all.

![Null Nullable Value Type](https://user-images.githubusercontent.com/26486200/126913886-9d2deed5-c33d-4c1d-a4ac-ca314162b707.png)
![Non-null value of Nullable Value Type](https://user-images.githubusercontent.com/26486200/126913911-74e0ec11-388e-44cc-90b9-70f4c6d244ae.png)
6. Some cases enforce annotation contract just fine. For instance 
```c#
static void NotNull_RefNullableAnnotated ( [NotNull] ref string? s )
{
} // Warning CS8777 Parameter 's' must have a non-null value when exiting..
```
7. While the same exhibits senseless warnings
```c#
static void NotNull_NullableAnnotated ( [NotNull] string? s )
{
} // Warning CS8777 Parameter 's' must have a non-null value when exiting.
```
8. Generally previous sample denotes ability to use contract attribute for variables that do not participate in it, i.e. they are not `by ref`.

```c#
static bool MaybeNullWhen__NullableAnnotated ( [MaybeNullWhen (true)] string s )
// No problem that s does not point to argument.
{
  if ( DateTime.Now.Ticks % 3 == 0 )
  {
    s = null;
    return true;
  }

  s = null; // No problem with null.
  return false;
}
```
9. Some are broken 
```c#
[return: NotNullIfNotNull ("a")]
[return: NotNullIfNotNull ("b")]
static string? NotNullIfNotNull ( string? a, string? b )
{
  if ( a is not null || b is not null )
    return null; // No problem here.
  
  return null;
}
```
…

**Comment by CyrusNajmabadi** [comment]
> Why it is null-state analysis if it works with nullable value types which do not have null value but default value?

I'm not sure what this means tbh.  Nullable value types definitely have a null value.  That's the reason they're called 'nullable' :-)

**Reply by deep-outcome** [reply]
They definitely does not have `null` value. [`public struct Nullable<T> where T : struct`](https://docs.microsoft.com/en-us/dotnet/api/system.nullable-1?view=net-5.0). `Nullable` is really unfortunate name. Only one true thing it expresses is that one can firmly use syntax `= null` that in fact does `= default(Nullable<>)` or `if(x is null)` but that is all. `Nullable<>` is `struct` thus it cannot be `null`. Never.

Method
```c#
public void Test ()
{
  int? i;
  i = null;
  i = new Nullable<int> ();
  i = new int? ();
  i = default (int?);
  i = default (Nullable<int>);
  i = default;
}
```
is in IL 

```csharp
.method /* 06000028 */ public hidebysig 
	instance void Test () cil managed 
{	
	.maxstack 1
	.locals /* 11000002 */ init (
		[0] valuetype [System.Runtime]System.Nullable`1<int32> i
	)
	
	IL_0000: nop
	// int? num = null;	
	IL_0001: ldloca.s 0
	IL_0003: initobj valuetype [System.Runtime]System.Nullable`1<int32> /* 1B000001 */
	// num = null;
	IL_0009: ldloca.s 0
	IL_000b: initobj valuetype [System.Runtime]System.Nullable`1<int32> /* 1B000001 */
	// num = null;
	IL_0011: ldloca.s 0
	IL_0013: initobj valuetype [System.Runtime]System.Nullable`1<int32> /* 1B000001 */
	// num = null;
	IL_0019: ldloca.s 0
	IL_001b: initobj valuetype [System.Runtime]System.Nullable`1<int32> /* 1B000001 */
	// num = null;
	IL_0021: ldloca.s 0
	IL_0023: initobj valuetype [System.Runtime]System.Nullable`1<int32> /* 1B000001 */
	// num = null;
	IL_0029: ldloca.s 0
	IL_002b: initobj valuetype [System.Runtime]System.Nullable`1<int32> /* 1B000001 */

	IL_0031: ret
}
```
Under any circumstance `i` is assigned with `value` not `null`.

**Reply by yaakov-h** [reply]
It depends how you define null.

https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUHgSymAH4ACKCAGyoEMAjKgUzIF4y4mAzW64AbjypMATgAUAIgC2ATwrU6jFgQDOZAN4AGAL4SYZAErcmYJlADGTAKIBHCLSoqxlGg2b6XVAJRlyEzxJkIGQStGQAbg4QLMAyAA5MEl78QA=

**Reply by deep-outcome** [reply]
@yaakov-h, I define nothing. `null` is technical blankly, exactly defined term specifying that reserved memory space of `nullable` `reference type` is whole zeroed. Unfortunate syntactic sugars or fancy but inaccurate naming cannot change this.

Your example just converts `null` into `default` under the hood.

**Reply by yaakov-h** [reply]
it does not.

**Reply by CyrusNajmabadi** [reply]
> They definitely does not have null value. 

They definitely do. 

You are incorrect here. 

```c#
Object o = null;
int? i = (int?)o;
```

They have a null value.

**Reply by CyrusNajmabadi** [reply]
> but that is all. Nullable<> is struct thus it cannot be null

This is incorrect. Definitionally. You can try this trivially both in the language, or in the runtime. Just use reflection and you will both read and write actual null values from nullable-value-types. Seriously, go try it! :-)

**Reply by CyrusNajmabadi** [reply]
Here you go:

```c#
using System;

class C
{
    public static int? i = 10;

    static void Main(string[] args)
    {
        var field = typeof(C).GetField(nameof(i));
        field.SetValue(null, null);
        var value = field.GetValue(null);
        Console.WriteLine(value == null);
    }
}
```

It *is* null. It's not some special struct value.  The runtime and language both agree on this fully.  Other trivial ways to see this:

```c#
        int? i = default;
        object o = i;
        Console.WriteLine(o == null);
```

If it was not null, you'd have a boxed non-null value that woudl not referece compare `==` to `null`.  But you do.  Because it's *literally* `null`.  Through the language, or directly just working with the runtime (with reflection) thisis all consistent around this area.

**Reply by CyrusNajmabadi** [reply]
```
	IL_0010: ldloc.1
	IL_0011: ldnull
	IL_0012: ceq
```

It's `null`.  Unless one wants to argue that `ldnull` and `ceq` are not actually comparing against null.

**Reply by svick** [reply]
@Uzivatel919 The way I would phrase it is that for reference types, `null` is the "null reference" (or "null pointer"). For nullable value types, `null` is *not* the null reference, but it is still `null`. And it is boxed to the null reference.

**Reply by deep-outcome** [reply]
> it does not.

You are right. You example violates [CA2013: Do not use ReferenceEquals with value types](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2013).

You example likely was about to somehow underline that in `Nullable<int>` struct was hidden `null` that when `boxed` to `object` was uncovered by smarty developer. But this is keen fairy tail story. Check with [Jon Skeet’s answer](https://stackoverflow.com/questions/3775582/how-is-the-boxing-unboxing-behavior-of-nullablet-possible#answer-3775643) for instance.

**Reply by deep-outcome** [reply]
> > They definitely does not have null value.
> 
> They definitely do.
> 
> You are incorrect here.
> 
> ```cs
> Object o = null;
> int? i = (int?)o;
> ```
> 
> They have a null value.

Likely I see what terminology is hidden among this. There is some realm pillar called `null` that in fact does not exists at all. It is _abstract keyword_ that any kind of realm inhabitant explains by its nature. Thus:

-  `reference type` is `null` when its memory space is all filled with zeros,
-  `value type` is never `null` since at least it is `default` instead,
- `Nullable<>` struct can be `null`; its `HasValue` `bool` property is false. This happens in well-known cases.
- If there come any new inhabitant that will be about to tell anyone it cannot provide useful value information, independently of its nature it will be presentable as `null` then.

**Reply by deep-outcome** [reply]
> > but that is all. Nullable<> is struct thus it cannot be null
> 
> This is incorrect. Definitionally. You can try this trivially both in the language, or in the runtime. Just use reflection and you will both read and write actual null values from nullable-value-types. Seriously, go try it! :-)

I do it day by day. Just proper understanding was passing me aside. I saw it just as syntactic sugar (such said) but now I know it has also full semantic meaning.

**Reply by CyrusNajmabadi** [reply]
> But this is keen fairy tail story.

There is no fairy tail here. The reflection examples demonstrate this. Again, I literally was part of the effort that did this work back in c#2. It's one of the first things I worked on in this group at ms after I moved over to devdiv :-)

**Reply by CyrusNajmabadi** [reply]
> There is some realm pillar called null that in fact does not exists at all. It is abstract keyword that any kind of realm inhabitant explains by its nature

So I'm not sure how to evaluate this statement. `null` definitely exists and has a definition.  Yes, you can say it has a description behind it... But that's true of everything in the language. There is a specification that states the semantics it should have and what is legal for it or not. 

However, there is no different sort of 'null' in c# or .net.  It's the same null at the language level and the runtime level. You can see that with my IL example.  You can see the simplest opcodes for referring to null and then comparing against it.  It's not something different. It's just null.

**Reply by deep-outcome** [reply]
> It's just null.

Ok, ok. Everything `nullable` has its own `null` state, `null` equality, …

**Comment by CyrusNajmabadi** [comment]
> Why they use null keyword when in .NET environment is preferable to use default whenever possible. 

I know of no such preference in .net.  The preference is to use 'default' for things that may not be null (for example value types, or unconstrained generics). Did things that can be null, 'null' is perfectly fine to use

**Reply by deep-outcome** [reply]
I can tell theorem right now but I am 100 % confident that `default` is suggested for any case. It is for instance more comprehensive. `Null` can appear more declarative though. For instance for `NotNull` attribute it was more beneficial if it would be `NotDefault` since that is the case.

**Reply by 333fred** [reply]
> I can tell theorem right now but I am 100 % confident that `default` is suggested for any case.

Certainly not. In Roslyn, we actually have an analyzer that enforces using `null`, not `default`, whenever possible.

**Reply by deep-outcome** [reply]
@333fred, hmm, maybe it is some abandoned recommendation. I was never stuck with it. I am almost sure it was presented as beneficial. I can image case when `class` is change to `struct` right now. But this could be debatable, of course. On the other hand for `NotNull`/`NotDefault` attribute it would be exact case. Passing around claim that `struct` can be `null` is unfortunate. It can be checked against `null`. But not `null`.

**Reply by CyrusNajmabadi** [reply]
> But not null.

You are incorrect. This is how it is literally defined in the language. *and* even if you don't care about the language, this is how the runtime works (which is trivial to verify).

**Reply by CyrusNajmabadi** [reply]
> hmm, maybe it is some abandoned recommendation. 

I don't think any such recommendation has ever existed. We have been firm and clear on what null is and how it relates to nullable value types since c# 2.

**Reply by 333fred** [reply]
> @333fred, hmm, maybe it is some abandoned recommendation.

Definitely not abandoned. This is very intentional on our part, and it's certainly my recommendation to any other user. `null` vs `default` gives important information: whether the value is some default for the domain (such as `default(int)`), or whether it's actually the absence of information, as `null` is.

**Reply by CyrusNajmabadi** [reply]
I can see someone using `default` for all use cases. It's consistent and simple. However, I'm with Fred that I prefer `null` for cases where `null` is the value. It's shorter, provides useful information, and is idiomatic.

**Reply by deep-outcome** [reply]
> > But not null.
> 
> You are incorrect. This is how it is literally defined in the language. _and_ even if you don't care about the language, this is how the runtime works (which is trivial to verify).

I care about language. I mean I like it. But. Can you provide me with some reference?

I just went through _ECMA-335 6th Edition / June 2012_ finding just indirect clause.

>Some methods traffic in “default values”. For a reference type, the default value is null; for a
nullable value type, the default value is HasValue returns false; for a non-nullable value type, the
default value is all-bits-zero (which for Boolean represents false, and for all arithmetic types
represents zero).

**Comment by CyrusNajmabadi** [comment]
> Thus to state out that null is fine value for instance of Nullable<int> variable is not true at all.

I genuinely cannot find where on the net this quoted statement comes from. However, I firmly disagree with it. `null` is absolutely fine for *any* nullable value type. That's literally the point of that feature. If there are docs or statements to effect that day otherwise, we should fix them.

**Reply by deep-outcome** [reply]
It cites me from another post.

>  That's literally the point of that feature. 

No it is not. The point is not to be sure that some `value type` _is / is not_ `null` but to be sure whether it _has / has not_ value. I was never curious if my `Nullable<XYZ>` has `null` value (which cannot have). Each time I was just curious if it `HasValue`.

```c#
public void Test2 ()
{
  int? i = null;
  if ( i.HasValue ) { }
  if ( i is not null ) { }
  if ( i != null ) { }
}
```

in IL

```csharp
.method /* 06000029 */ public hidebysig 
	instance void Test2 () cil managed 
{
	.maxstack 1
	.locals /* 11000009 */ init (
		[0] valuetype [System.Runtime]System.Nullable`1<int32> i,
		[1] bool,
		[2] bool,
		[3] bool
	)
	IL_0000: nop
	// int? num = null;

	IL_0001: ldloca.s 0
	IL_0003: initobj valuetype [System.Runtime]System.Nullable`1<int32> /* 1B000001 */
	// if (num.HasValue)

	IL_0009: ldloca.s 0
	IL_000b: call instance bool valuetype [System.Runtime]System.Nullable`1<int32>::get_HasValue() /* 0A000017 */
	IL_0010: stloc.1
	// sequence point: hidden
	IL_0011: ldloc.1
	IL_0012: brfalse.s IL_0016

	IL_0016: ldloca.s 0
	IL_0018: call instance bool valuetype [System.Runtime]System.Nullable`1<int32>::get_HasValue() /* 0A000017 */
	IL_001d: stloc.2
	// sequence point: hidden
	IL_001e: ldloc.2
	IL_001f: brfalse.s IL_0023

	IL_0023: ldloca.s 0
	IL_0025: call instance bool valuetype [System.Runtime]System.Nullable`1<int32>::get_HasValue() /* 0A000017 */
	IL_002a: stloc.3

	IL_002b: ldloc.3
	IL_002c: brfalse.s IL_0030

	IL_0030: ret
}
```

`HasValue` all the time,

---
I believe that is the very fundamental piece that `null` value is strictly dedicated to `reference types`. Am I wrong?

**Reply by CyrusNajmabadi** [reply]
> Am I wrong?

Yes. You are wrong.  I say this as one of the people who worked on the design and impl of Nullable<T>.

**Reply by CyrusNajmabadi** [reply]
> It cites me from another post.

Sorry, citing yourself isn't really an argument :). Your position does not match the language or runtime spec or implementation.

**Reply by HaloFour** [reply]
`Nullable<T>` isn't a normal struct.  The runtime has special knowledge of it and will implicitly convert between it and a `null` reference.  For example:

```cs
int? i = null;
object o = i;

Debug.Assert(o == null);
```

If this weren't the case `o` would be a boxed copy of a `Nullable<T>` struct that contained the sentinel value representing `null`.   But that's not the case, `o` is an actual `null` reference.

**Reply by deep-outcome** [reply]
> > It cites me from another post.
> 
> Sorry, citing yourself isn't really an argument :). Your position does not match the language or runtime spec or implementation.

It was not about to present any meaning. That is obvious, of course.

**Reply by deep-outcome** [reply]
> > Am I wrong?
> 
> Yes. You are wrong. I say this as one of the people who worked on the design and impl of Nullable.

I see now. I was never aware that `null` and `null reference` are not interchangeable terms. Seeing that `null` is just “colloquial” form of latter one. Now it is everything clear. `Null` can be this and that.

**Reply by CyrusNajmabadi** [reply]
`null` is just null. It's valid for (and 100% interchangeable) between direct reference types, and nullable value types. That's literally why they're called 'nullable'. Because the language and runtime allow the exact same `null` for them that we've had since v1. 

Heck NVTs came in v2. `null` having this meaning had been in both if these places since almost the beginning. We literally added it to the system at the same time we did generics. They're baked in as deeply as we did that.

**Comment by khm1600** [comment]
> While the same exhibits senseless warnings
> ```cs
> static void NotNull_NullableAnnotated ( [NotNull] string? s )
> {
> } // Warning CS8777 Parameter 's' must have a non-null value when exiting.
> ```

[NotNullAttribute Class](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.notnullattribute?view=net-5.0):
> Specifies that an _input_ argument was not `null` when the call _returns_.

In other words, when the arg is `null`, the call does not return. Therefore:

```cs
static void NotNull_NullableAnnotated([NotNull] string? s)
{
    if (s is null) throw new Exception();
} // no CS8777
```

**Reply by jnm2** [reply]
@Uzivatel919 Yes, what's senseless is not the warning. What's senseless is putting that attribute on that parameter.

**Reply by Joe4evr** [reply]
Is it senseless? [The new `ANE.ThrowIfNull` helper method is implemented in exactly that way.](https://github.com/dotnet/runtime/pull/55594/files#diff-1992d21061328d688e1eb3d2e734a59c96a41335801e0e51cacabfa3f1043b71R59)

**Reply by deep-outcome** [reply]
@jnm2, in fact I was expecting error for putting the attribute to not `by ref` `parameter`. This is not there. Instead there is senseless warning. On the other I found it out by mistake as you took hint.

**Reply by deep-outcome** [reply]
@Joe4evr, No, it is not. I missed the conceptuality. I got impression that these ones are designed to denotes contract about outputs but they are leveraged also to denotes contract about inputs. Handy.

**Reply by jnm2** [reply]
@Joe4evr You're right, and it's strange that I forgot that given that I've made use of it when annotating two assertion frameworks.
